### DISCUSSION TITLE
Input/output signals ESP32 A2DP

### DISCUSSION THREAD

**Discussion by JohanWellander** [pschatzmann/arduino-audio-tools#1055] [Q&A] [answered]
Hello Phil!

I am currently working on a project where I will stream music via Bluetooth using ESP32 and I found the great work you have done regarding this. I'm new to signal processing and have a question about how it works.

My plan is to:
1. Receive music via Bluetooth to an ESP32 from a phone or a computer. 
2. Output the music through one (or several) of the pins. 
3. The output from the ESP32 should then be an input to a second ESP32 through one (or several) of the pins.
4. Finally I want to transmit the music from the second ESP32 via Bluetooth to an external Bluetooth speaker.  

So my question is, do you have any tips or tricks on how I can do this? I have managed to receive the signals from my phone and also transmit sound to a Bluetooth speaker. The problem I encounter is in step 2 and 3 since I'm not sure of what kind of signals the output is or what it should be, for example, I2S or analog? Further on I'm not sure how to handle the input to the second ESP32, what kind of signal is coming in, and if I need to process the signal to transmit it to a Bluetooth speaker?

**Comment by pschatzmann** [comment, marked as answer]
First of all, I don't see the added value of doing this: you can connect directly from your phone to the Bluetooth speaker.
If you still want to do this you have quite a few challenges. But you can find examples for all the involved steps in this project!

The biggest challenge that I see is to transmit the signal between the 2 ESPs reliably. 

I see the following options:

- I2S (I don't think this will work reliably, most people report lost samples)
- Built in DAC -> ADC: I expect the quality to be not good, though some users were quite happy with this. But you can forget using proper stereo, since ADC supports only one channel.
- Serial connection using some compressed data with the help of a codec (e.g ADPCM)

I suggest to prototype all 3 options: I would tend to go for the third one.
In any case A2DP is providing and [using PCM data as input](https://github.com/pschatzmann/ESP32-A2DP#sending-data-from-a-a2ds-data-source-with-a-callback).

ps. I suggest to go thru the Wiki of this project..
